evm_arithmetization/src/verifier.rs

```diff
@@ -89,328 +77,374 @@ pub(crate) fn initial_memory_merkle_cap<
+    use ethereum_types::{BigEndianHash, U256};
+    use itertools::Itertools;
+    use plonky2::field::extension::Extendable;
+    use plonky2::hash::hash_types::RichField;
+    use plonky2::plonk::config::{GenericConfig, GenericHashOut};
+    use starky::config::StarkConfig;
+    use starky::cross_table_lookup::{get_ctl_vars_from_proofs, verify_cross_table_lookups};
+    use starky::lookup::GrandProductChallenge;
+    use starky::stark::Stark;
+    use starky::verifier::verify_stark_proof_with_challenges;
+
+    use crate::all_stark::Table;
+    use crate::cpu::kernel::aggregator::KERNEL;
+    use crate::cpu::kernel::constants::global_metadata::GlobalMetadata;
+    use crate::get_challenges::testing::AllProofChallenges;
+    use crate::memory::segments::Segment;
+    use crate::memory::VALUE_LIMBS;
+    use crate::proof::PublicValues;
+    use crate::proof::*;
+    use crate::util::h2u;
+    use crate::verifier::initial_memory_merkle_cap;
+    use crate::{AllStark, NUM_TABLES};
+
+    impl<F: RichField + Extendable<D>, const D: usize> AllStark<F, D> {
+        pub(crate) fn num_lookups_helper_columns(
+            &self,
+            config: &StarkConfig,
+        ) -> [usize; NUM_TABLES] {
+            [
+                self.arithmetic_stark.num_lookup_helper_columns(config),
+                self.byte_packing_stark.num_lookup_helper_columns(config),
+                self.cpu_stark.num_lookup_helper_columns(config),
+                self.keccak_stark.num_lookup_helper_columns(config),
+                self.keccak_sponge_stark.num_lookup_helper_columns(config),
+                self.logic_stark.num_lookup_helper_columns(config),
+                self.memory_stark.num_lookup_helper_columns(config),
+                self.mem_before_stark.num_lookup_helper_columns(config),
+                self.mem_after_stark.num_lookup_helper_columns(config),
+                #[cfg(feature = "cdk_erigon")]
+                self.poseidon_stark.num_lookup_helper_columns(config),
+            ]
         }
     }
 
-    Ok(())
-}
+    fn verify_initial_memory<
+        F: RichField + Extendable<D>,
+        C: GenericConfig<D, F = F>,
+        const D: usize,
+    >(
+        public_values: &PublicValues<F>,
+        config: &StarkConfig,
+    ) -> Result<()> {
+        for (hash1, hash2) in initial_memory_merkle_cap::<F, C, D>(
+            config.fri_config.rate_bits,
+            config.fri_config.cap_height,
+        )
+        .0
+        .iter()
+        .zip(public_values.mem_before.mem_cap.iter())
+        {
+            for (&limb1, limb2) in hash1.to_vec().iter().zip(hash2) {
+                ensure!(
+                    limb1 == F::from_canonical_u64(limb2.as_u64()),
+                    anyhow::Error::msg("Invalid initial MemBefore Merkle cap.")
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn verify_proof<
+        F: RichField + Extendable<D>,
+        C: GenericConfig<D, F = F>,
+        const D: usize,
```

Comment:
In that case we could gate behind a testing module the intermediate `verify_foo` methods in `fixed_recursive_verifier` too (i.e. `verify_root`. / `verify_segment_aggregation` / `verify_batch_aggregation`), as well as `prove_all_segments`

Reply:
Thanks for the catches! Done.